### DISCUSSION TITLE
HW-MS03 Microwave sensor need help with syntax while/else

### DISCUSSION THREAD

**Discussion by marinehm** [micropython/micropython#10571] [RP2040 / Pico]
Having trouble with While Else statement

```py
# MicroPython v1.19.1 on 2022-06-18; Raspberry Pi Pico with RP2040
# Thonny 4.0.2
# HW-MS03 Microwave motion sensor
# Created by Nick Sebring
# 01-24-2023
##################################################################
# IMPORT NECESSARY MODULES
from machine import Pin
from time import sleep
import utime
# ASSIGN PINS
led = Pin(6, Pin.OUT)
hw = Pin(8, Pin.IN)
`# MAKE THINGS HAPPEN`
while hw==0:
#     hw==0:  #light stays on regardless
#     hw==1:  #light stays on regardless
#     hw!=0:  #prints "Scaning..." FOREVER!
#     hw!=1:  #prints "Scaning..." FOREVER!
  led.off()
  print("Scanning...")
  sleep(5)  #gives me time to figure what's going on
else:
  led.on()
  print("Motion detected!")
  sleep(5)   #gives me time to figure what's going on  
#What I expect; if after 5 seconds of no motion, the led to turn off
#But, if the motion detector remains high, then the led will remain on
#until the sensor goes LOW (0)'''

**Comment by davefes** [comment]
`my_pin = hw.value()`
Then test my_pin

`while my_pin == 0:`

**Comment by robert-hh** [comment]
@marinehm About formatting: When adding code, either enclose the code in lines containing three backticks ``` or indent the code lines by like four spaces. An easy way to get the 3 backtick lines is to fully highlight the pasted code and push Ctrl-E.

**Comment by sosi-deadeye** [comment]
I'm uncertain if this works (not tested).
You could use a combination of `Pin.irq` and `Timer`.

Here the code:
```python
from machine import Pin
from machine import Timer


class Motion:
    def __init__(self, motion_pin, led_pin, off_delay):
        # motion_pin, not motion
        self.motion = Pin(motion_pin, mode=Pin.IN)
        self.led = Pin(led_pin, mode=Pin.OUT, value=0)
        self.off_delay = off_delay
        self.timer = None
        self.motion.irq(
            handler=self._callback,
            trigger=Pin.IRQ_FALLING | Pin.IRQ_RISING,
        )
    
    def _callback(self, pin):
        if self.timer:
            self.timer.deinit()
            self.timer = None
            
        if pin.value():
            self.led.value(1)
        else:
            # This should start the Timer
            # if your port does not support software timers,
            # then use a positive ID for initialization 
            self.timer = Timer(
                -1,
                mode=Timer.ONE_SHOT,
                period=self.off_delay,
                callback=self._no_motion,
            )
    
    def _no_motion(self, timer):
        self.led.value(0)



motion1 = Motion(motion_pin=8, led_pin=6, off_delay=5_000)
```

The `irq` watches both edges (`falling` and `rising`). I guess if motion is detected, the `self.led` should turn on. A `rising edge` is an input change from `0` to `1`. This should switch the `self.led` on. The `falling edge` should turn the `self.led` off after 5 seconds (`5_000` ms). A ONE_SHOT `Timer` with a callback is started. So if the input changes from 1 to 0, then the `Timer` is created, which calls after 5 seconds the `_no_motion` method. If during this `period` a `rising edge` happens, the `self.timer.deinit()` is called, to stop the `Timer` object. Then `None` is assigned to `self.timer`. I do not know, if you can reuse `Timers`. I think not.

**Reply by marinehm** [reply]
This is what I get when run:
```Traceback (most recent call last):
  File "<stdin>", line 39, in <module>
  File "<stdin>", line 7, in __init__
NameError: name 'motion' isn't defined

**Reply by sosi-deadeye** [reply]
Easy fix
```python
# change
self.motion = Pin(motion, mode=Pin.IN)
# to
self.motion = Pin(motion_pin, mode=Pin.IN)
```

**Reply by marinehm** [reply]
The longer the wire, the more "collective?", the light stays on.....  More investigation!

**Comment by marinehm** [comment]
```py
# MicroPython v1.19.1 on 2022-06-18; Raspberry Pi Pico with RP2040
# Thonny 4.0.2
# HW-MS03 Microwave motion sensor
# Created by Nick Sebring
# 01-24-2023
##################################################################

# IMPORT NECESSARY MODULES
from machine import Pin
from time import sleep
import utime

# ASSIGN PINS
led = Pin(6, Pin.OUT)
hw = Pin(8, Pin.IN)
my_pin = hw.value()

# MAKE THINGS HAPPEN
while True:
  if my_pin == 0:
#     my_pin ==0:    prints "Scanning..." FOREVER!
#     my_pin !=0:    prints "Motion detected!" FOREVER!
#     my_pin == 1:   prints "Scanning..." FOREVER!
#     my_pin != 1:   prints "Motion detected!" FOREVER!
    led.off()
    print("Scanning...")
    sleep(5)   #gives me time to figure what's going on
  else:
    led.on()
    print("Motion detected!")
    sleep(5)   #gives me time to figure what's going on
```
I just noticed; each time I run it, it toggles whether the led is on or off.  WHAT THE HECK!?
Not like what is commented out.

**Reply by karfas** [reply]
What do you expect ?

You are setting my_pin to the input value and don't change it during the while True loop.

**Comment by marinehm** [comment]
I like the ctrl E
I'm used to the <> from the Arduino forum,  paste the code between < code here >
**OK, got it thanks!**

**Comment by peterhinch** [comment]
As a general point it is rarely necessary to write `while else`. Referring to the above code, the following behaves identically:
```python
while hw == 0:
    # do stuff
led.on()  # No need for else
print("motion detected")
```
The `else` clause is rather specialised. It runs if the `while` loop exits normally. It does not run if it exits via `break`, `return` or via an exception being thrown.

**Reply by davefes** [reply]
Suggestion:  use 4 spaces to indent.

Also, I would either define your pin as a PULL_UP or put a 10K resistor from it to 3V3 to properly define it's operating state.

**Reply by marinehm** [reply]
I do use a resister on all my LEDs.

**Comment by bixb922** [comment]
If you want to test the input pin every 5 seconds, you need to test hw.value() within the while loop. You are testing the pin only once at the beginning before the loop, is that as intended?

**Comment by marinehm** [comment]
Copying my Arduino sketch that works flawlessly.  I just want to learn micropython and play with some new Raspberry Pi Picos.

```
/*
Microwave Motion Sensor
LOLIN (WEMOS) Mini-D Pro
HW-MS03 Microwave motion sensor
1000uf 25V Capacitor
Lolin Relay

Capacitor across the  +  -

Senses motion from front and back of the PCB.  Not so much from the sides.
This even worked through a fairly thick plastic table that I used as a barrier.

By Nick Senring; 12-16-2022
*/

int HWPin = 16;
int Relay = 14;
int val;                          // This is the value from the HW-MS03 Microwave motion sensor

void setup()
{
  Serial.begin(9600);
  pinMode(HWPin, INPUT);          // HW-MS03 Microwave motion sensor
  pinMode(BUILTIN_LED, OUTPUT);   // BLUE LED onboard as OUTPUT (High/LOW is inverted)
  pinMode(Relay, OUTPUT);
}
void loop()
{
  val = digitalRead(HWPin);
  //low = no motion, high = motion
  if (val == HIGH)
  {
    Serial.println("Motion detected.  ");
    digitalWrite(BUILTIN_LED, LOW);
    digitalWrite(Relay, HIGH);  //Turn something on
    delay(30000); //   5000= 5 sec   300000= 5min, 600000= 10 min
    //  Keep it on for a period of time  //
  }

  else
  {
    Serial.println("Scanning...");
    digitalWrite(BUILTIN_LED, HIGH);
    digitalWrite(Relay, LOW);  //Turn something off
  }
}
```

**Comment by marinehm** [comment]
This doesn't work either:
```
# MicroPython v1.19.1 on 2022-06-18; Raspberry Pi Pico with RP2040
# Thonny 4.0.2
# HW-MS03 Microwave motion sensor
# Created by Nick Sebring
# 01-24-2023
##################################################################

# IMPORT NECESSARY MODULES
from machine import Pin
from time import sleep
import utime

# ASSIGN PINS
led = Pin(6, Pin.OUT)
hw = Pin(8, Pin.IN, Pin.PULL_DOWN)
#my_pin = hw.value()
led.off()


# MAKE THINGS HAPPEN
while True:
  while hw==high: # IS THIS HOW TO CHECK THE HW PIN(8) HIGH/LOW?   **Similar to "digital read" in Arduino?**
    # do stuff
    led.on()  # No need for else
    print("motion detected")
    sleep(5) # So I can see what's going on until it works correctly

    led.off()  # HOW ELSE IS THE LIGHT GOING TO TURN OFF?
    print("Scanning...")
    sleep(2)
```

**Reply by davefes** [reply]
Try:
```py
# IMPORT NECESSARY MODULES
from machine import Pin
from time import sleep
import utime

# ASSIGN PINS
led = Pin(6, Pin.OUT)
hw = Pin(8, Pin.IN, Pin.PULL_DOWN)
led.off()


# MAKE THINGS HAPPEN
while True:
  my_pin = hw.value()

  while my_pin: #  if the pin is high do this
    # do stuff
    led.on()  # No need for else
    print("motion detected")
    sleep(5) # So I can see what's going on until it works correctly

    led.off()  # HOW ELSE IS THE LIGHT GOING TO TURN OFF?
    print("Scanning...")
    sleep(2)

    break
```
Haven't checked it.

**Reply by robert-hh** [reply]
Looking at the code, once the inner while loop is entered, it will not be left, because the value of my_pin does not change. Is that intended?

**Reply by davefes** [reply]
Nope!  Added a break.  I was too focused on "getting into the loop".

**Reply by marinehm** [reply]
That code just toggled the LED on/off.  for the amount of time on the "sleep".

THE Microwave sensor is either 0 or 1  rather; **high** / **low**.

Not sure how to "read" the Pin in Mycropython to get the value from Pin 8
Would this need an ADC?

No.  I only want the LED to come on when the Sensor detects motion and set the pin high.  Stay on for a specified amount of time.  If it is still high, will run again for the specified time until it no longer detects motion.  When the sensor goes **low** and the code runs again after the sleep, I want the LED to be turned off.  

Please see my Arduino code above.  It works flawlessly  So, why am I doing this?  To learn Micropython.

**Reply by marinehm** [reply]
So I did this:
It works but, I pretty much have to touch the sensor.  Unlike on my ESP8266 with Arduino.IDE, it senses from over 7 feet and through my plastic folding table.  What could make that difference?
And if you noticed...  I changed LED pin to 3.  Pin 6 stopped working.  Afterward, I think with my failing eyesight, I plugged the wire into the wrong hole.    Hahaha  **That's what she said!**  3 is working
```
# IMPORT NECESSARY MODULES
from machine import Pin
from time import sleep
import utime

# ASSIGN PINS
led = Pin(3, Pin.OUT)
hw = Pin(8, Pin.IN, Pin.PULL_DOWN)
my_pin = hw.value()

led.off()
# MAKE THINGS HAPPEN
while True:
  my_pin = hw.value()

  while my_pin==hw.value(): #  if the pin is high do this
    # do stuff
    if my_pin > 0:
        led.on()  # No need for else
        print("Motion detected!")
        sleep(2) # So I can see what's going on until it works correctly
    else:
        led.off()  # HOW ELSE IS THE LIGHT GOING TO TURN OFF?
        print("Scanning...")
        sleep(2)

    break
```

**Reply by karfas** [reply]
```
while True:
  my_pin = hw.value()
  while my_pin==hw.value(): #  if the pin is high do this
    break
```
This is a creative (but strange) way to use while...break. You are using `while` as replacement for an `if`.

Also note that `while my_pin==hw.value():` will not work as stated in the comment..
The code inside the `while` will run when the input **does not change** in respect to the value read in the `while True`  loop.
This is different to "if the pin is high" !

In my opinion, you can reduce all this to 
```
while True:
    my_pin = hw.value()
    if my_pin > 0:
        led.on()  # No need for else
        print("Motion detected!")
        sleep(2) # So I can see what's going on until it works correctly
    else:
        led.off()  # HOW ELSE IS THE LIGHT GOING TO TURN OFF?
        print("Scanning...")
        sleep(2)

**Reply by marinehm** [reply]
This makes more sense than my code.  It does work and the code is "clean".  Now to figure out why I have to touch the sensor for it to notice "motion".

**Comment by bixb922** [comment]
I don't see the definition of "high".
 I suppose you want to check for a value of 1 on pin 8. The variable hw has the pin object, not the pin value. You are comparing the pin object with 1 or 0,  that will always be false. The pin value is hw.value()

In case of doubt, use print to get what is happening.

**Reply by davefes** [reply]
For a ESP32:

```py
hw = Pin(32, Pin.IN, Pin.PULL_DOWN)
my_pin = hw.value()
```
`print (my_pin)
`
and it should read 0  because it is pulled-down and I expected that the code would not run.

When you take pin8 hi then it should read 1 and 
the rest of the code should run

Which works for me.

**Reply by davefes** [reply]
Sorry, I have been working on a ESP32 and pin 8 is not allowed:
https://github.com/peterhinch/micropython-samples/tree/master/ESP32
see the .pdf

Are there pins on your device that one shouldn't use?

**Comment by marinehm** [comment]
In the REPL:  if I type: 
`print(hw())`
I get the value of the hw Pin .  How do I ask if it's either high or low?
When I import Pin and then help(Pin), I see that "high" and "low" are valid arguments.

**Reply by robert-hh** [reply]
high() and low() are valid methods (function calls), not names for returned values. a call to hw() or hw.value() returns 0 or 1, depending on the input level. 0 maps to a low level, 1 to a high level.

**Reply by bixb922** [reply]
hw.value()==1 to test if on, hw.value()==0 to test if off.

**Comment by marinehm** [comment]
Something strange just happened.  I was pulling off the sensor to hook it back up to my ESP8266 and run the Arduino sketch on it.  The LED came on without the sensor connected.  Playing around with this some more.  The jumper wire still connected to the Pico HW Pin.  It acts as a capacitance touch sensor.  Senses my finger at close to 2 inches from the wire and the LED comes on.  No Sensor!  This will work for me with turning on the light for my shed.  BUT, I wanted the motion detector for my hall and bathroom light.  For now, I will use the esp8266/Arduino.IDE until I investigate this phenomenon some more.

**Reply by peterhinch** [reply]
CMOS inputs are very high impedance. This behaviour is well known, but it can't really be relied on. There are also hazards with static discharge. Touching an open circuit input can damage the device if your body becomes electrically charged.

**Comment by marinehm** [comment]
I need some work with Micropython.
Here's my completed Arduino sketch:
```
/*
  Microwave Motion Sensor
  LOLIN (WEMOS) Mini-D Pro
  HW-MS03 Microwave motion sensor
  1000uf 25V Capacitor
  Lolin Relay
  MQTT to publish "Motion detected!" so I can log it from my MQTT Server

//  Capacitor across the  +  - was removed from the final product, will re-evaluate
//  when the relay is cycled on/off a few times.

  Senses motion from front and back of the PCB.  Not so much from the sides.
  This even worked through a fairly thick plastic table that I used as a barrier.

  By Nick Sebring; 2-4-2023
*/

// INSTALL WIFI, MQTT
#include <ESP8266WiFi.h>
#include <PubSubClient.h>

// Connect to the WiFi
const char* ssid = "MYSSID";
const char* password = "MYPASSWORD";
const char* mqtt_server = "192.168.1.13";

WiFiClient espClient;
PubSubClient client(espClient);

boolean reconnect() {  // **********************************************************
  // Loop until we're reconnected
  while (!client.connected()) {
    Serial.print (F("Contacting MQTT server..."));
    // Attempt to connect
    { if (client.connect("MCSHED"))    //assign a "client name".  Each wemos must have a unique name
        Serial.println (F("connected"));

      return client.connected();

      Serial.print (F("Failed to connect. "));
      Serial.println (F(" Attempting connection again in 3 seconds"));
      // Wait 3 seconds before retrying
      //      delay(3000);
      return 0;
    }
  }
}

void setup()
{
  {
    Serial.begin(9600);
    client.setServer(mqtt_server, 1883);
  }
  // Connect to WiFinetwork
  Serial.println();
  Serial.println();
  Serial.print (F("Connecting to "));
  Serial.println(ssid);
  WiFi.begin(ssid, password);
  WiFi.mode(WIFI_STA);
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    //Serial.begin(9600);
    Serial.print (F("."));
  }
  Serial.println (F(""));
  Serial.println (F("WiFi connected"));
  // Print the IP address
  Serial.print (F("Local IP: "));
  Serial.println(WiFi.localIP());

  //  *********************************************************************************

  Serial.println(F("Microwave sensor located in the Motorcycle shed"));


}
void loop()
{
  if (!client.connected()) {
    reconnect();
  }
  
  {
    int HWPin = 16;
    int Relay = 14;
    int val;                          // This is the value from the HW-MS03 Microwave motion sensor

    pinMode(HWPin, INPUT);          // HW-MS03 Microwave motion sensor
    pinMode(BUILTIN_LED, OUTPUT);   // BLUE LED onboard as OUTPUT (High/LOW is inverted)
    pinMode(Relay, OUTPUT);
    val = digitalRead(HWPin);

    if (val == HIGH)  //low = no motion, high = motion
    { Serial.println("Motion detected.  ");
      digitalWrite(BUILTIN_LED, LOW);
      digitalWrite(Relay, HIGH);  //Turn something on
      client.publish("Outside/MC_shed", "Motion detected!");
      delay(450000); //   5000= 5 sec   300000= 5min, 600000= 10 min
      //  Keep it on for a period of time  //
    }

    else
    {
      Serial.println("Scanning...");
      digitalWrite(BUILTIN_LED, HIGH);
      digitalWrite(Relay, LOW);  //Turn something off
    }
  }
}
```
